bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/internal/servlet/CameraServlet.java

```diff
@@ -0,0 +1,311 @@
+        if (req == null || resp == null) {
+            return;
+        }
+        String pathInfo = req.getPathInfo();
+        if (pathInfo == null) {
+            return;
+        }
+        logger.debug("GET:{}, received from {}", pathInfo, req.getRemoteHost());
+        if (!"DISABLE".equals(handler.getWhiteList())) {
+            String requestIP = "(" + req.getRemoteHost() + ")";
+            if (!handler.getWhiteList().contains(requestIP)) {
+                logger.warn("The request made from {} was not in the whiteList and will be ignored.", requestIP);
+                return;
+            }
+        }
+        switch (pathInfo) {
+            case "/ipcamera.m3u8":
+                Ffmpeg localFfmpeg = handler.ffmpegHLS;
+                if (localFfmpeg == null) {
+                    handler.setupFfmpegFormat(FFmpegFormat.HLS);
+                } else if (!localFfmpeg.getIsAlive()) {
+                    localFfmpeg.startConverting();
+                } else {
+                    localFfmpeg.setKeepAlive(8);
+                    sendFile(resp, pathInfo, "application/x-mpegURL");
+                    return;
+                }
+                // Allow files to be created, or you get old m3u8 from the last time this ran.
+                try {
+                    Thread.sleep(HLS_STARTUP_DELAY_MS);
+                } catch (InterruptedException e) {
+                }
+                sendFile(resp, pathInfo, "application/x-mpegURL");
+                return;
+            case "/ipcamera.mpd":
+                sendFile(resp, pathInfo, "application/dash+xml");
+                return;
+            case "/ipcamera.gif":
+                sendFile(resp, pathInfo, "image/gif");
+                return;
+            case "/ipcamera.jpg":
+                sendSnapshotImage(resp, "image/jpg");
+                return;
+            case "/snapshots.mjpeg":
+                req.getSession().setMaxInactiveInterval(0);
+                snapshotStreamsOpen++;
+                handler.streamingSnapshotMjpeg = true;
+                handler.startSnapshotPolling();
+                StreamOutput output = new StreamOutput(resp);
+                do {
+                    try {
+                        output.sendSnapshotBasedFrame(handler.getSnapshot());
+                        Thread.sleep(1005);
+                    } catch (InterruptedException | IOException e) {
+                        // Never stop streaming until IOException. Occurs when browser stops the stream.
+                        snapshotStreamsOpen--;
+                        if (snapshotStreamsOpen == 0) {
+                            handler.streamingSnapshotMjpeg = false;
+                            handler.stopSnapshotPolling();
+                            logger.debug("All snapshots.mjpeg streams have stopped.");
+                        }
+                        return;
+                    }
+                } while (true);
```

Comment:
If I understand correctly, this runs as long as the user has opened the browser. So, this will occupy a thread of the HTTP server. Are you aware from which thread pool you get the thread from? Same for the others.

Reply:
It will keep the thread until the stream is not accepted by a browser. You can press stop on a browser to halt it or close the tab. Also if someone has lazy loading on a widget setup it can open and stay open for a long time even when navigated away from the page that has the widget. To stop that it requires the refresh to be pressed when not on the widgets page.

I do not know where the threads are coming from only that each request will get its own dedicated thread hence why I feel it is ok to use thread sleep. Thank you for looking at this as it is not my strong point.

Reply:
It will probably consume a thread from the Jetty thread pool, which defaults to 50. https://github.com/openhab/openhab-distro/blob/9bdedaa6f537b8b31bfd4082ee77f5a9401f19b2/distributions/openhab/src/main/resources/userdata/etc/org.ops4j.pax.web.cfg#L9

I think that's ok.